Resolve sluggish field behaviour

chrisvxd opened the pull request:
Massively improve performance of field rendering by introduction of a separate `fieldStore`. This eliminates unnecessary re-renders when updating unrelated fields.

## Testing

* Test drag and drop of array fields
* Ensure undo/redo history restores array fields appropriately after moving things around
* Test complex combinations of fields (objects in arrays in arrays etc)

Closes #1429, closes #1450

FedericoBonel commented:
Some bugs I found so far:

When using the following config:

```tsx
export const conf = {
  components: {
    TestNested: {
      fields: {
        array: {
          type: "array",
          arrayFields: {
            title: { type: "text" },
            nestedArray: {
              type: "array",
              arrayFields: { title: { type: "text" } },
            },
          },
          defaultItemProps: {
            title: "Title",
            nestedArray: [],
          },
        },
      },
      defaultProps: {
        array: [],
      },
      render: ({ array }) => {
        return (
          <ul>
            {array.map((arrayItem, outIndex) => (
              <li key={outIndex}>
                {arrayItem.title}
                <ul>
                  {arrayItem.nestedArray.map((nestedItem, inIndex) => (
                    <li key={`${inIndex}`}>{nestedItem.title}</li>
                  ))}
                </ul>
              </li>
            ))}
          </ul>
        );
      },
    },
  },
};
```

### Nested items disappear 

1. Add three outer items.  
2. Add three nested items to one of them, and change its order.
4. Nested items disappear (they reappear if you add a new nested item).  

This does not happen if the items are collapsed.  

https://github.com/user-attachments/assets/de2640d9-c708-4c3f-92e6-4d65e08acaa2  

### Adds extra nested items

1. Add three outer items.  
2. Add three nested items to the first outer item.  
3. Add two nested items to the second outer item.  
4. Add one nested item to the third outer item.  
5. Expand the third outer item, leave all its nested items collapsed, then:  
   - Move it to the first position.  
   - Move it to the last position.  
   - Move it to the middle position.  
6. A nested item is added (not shown in the canvas) when moved to the first position.  
7. Another nested item is added when moved back to the third position.  
8. An item is removed when moved to the second position.  

https://github.com/user-attachments/assets/b8d61f1a-6f97-407b-beef-77dc8cda6e98

FedericoBonel commented:
Found another bug, it seems like `AutoFields` are not working correctly.

Using this config:

```tsx
const config = {
  TestCustom: {
    fields: {
      array: {
        type: "custom",
        render: ({ value, onChange }) => {
          return (
            <AutoField
              field={{
                type: "array",
                arrayFields: { title: { type: "text" } },
                defaultItemProps: { title: "Title" },
              }}
              onChange={onChange}
              value={value}
            />
          );
        },
      },
    },
    defaultProps: {
      array: [],
    },
    render: ({ array }) => {
      return (
        <ul>
          {array.map((item, index) => (
            <li key={index}>{item.title}</li>
          ))}
        </ul>
      );
    },
  },
};
```

1. Drag and drop a `TestCustom` component.  
2. Add items.  
3. Expand one of the items.  
   - The `title` field is empty even though the default item prop is visible in the canvas.  
   - Trying to change the `title` field updates it in the canvas, but as soon as you click away from the field, it resets to empty in the sidebar.  
4. Reorder one of the items.  
   - All items disappear from the fields.  

https://github.com/user-attachments/assets/a971ab23-0d5e-49cc-85ce-a15edd4aa690

dj-fiorex commented:
is there any news on this? we need performance improvements for array (and slots) 👍

FedericoBonel commented:
This issue might get closed with this PR: #1450

chrisvxd commented:
@FedericoBonel ready for re-review!

I rearchitected and simplified the approach:

- Introduced a new `fieldStore` that works for both internal and external AutoField use.
  - This field gets synchronised with the primary `appStore` when internal, and with the provided `value` when used with an external field.
  - This unifies most of the code paths between external and internal use, simplifying the implementation
  - This also means that external use of AutoField also benefits from much of the same performance improvements
- Also simplified the array field implementation, eliminating much of the duplicated state
- Updated against main, addressing RTE performance in array fields

I'm hoping that these changes simplify this enough to eliminate most remaining bugs and make any remaining ones easier to address. Please do your thing!

Drewbiee123 commented:
- If you ever need help reviewing similar edge cases, feel free to reach out

Sent from Yahoo Mail for iPhone


On Sunday, December 14, 2025, 11:54 PM, ***@***.*** wrote:

Thanks Federico — really appreciate you re-testing and confirming everything is fixed.Glad the array + object field behaviors are stable now (add/duplicate/move/focus/dispatcher/root object).I’m going to close this out on my side. If you hit any remaining edge case, please drop it here or open a new issue with a minimal repro and I’ll jump on it.


Sent from Yahoo Mail for iPhone


On Sunday, December 14, 2025, 11:47 PM, Federico Bonel Tozzi ***@***.***> wrote:

FedericoBonel left a comment (puckeditor/puck#1213)
Beautiful! Everything has been fixed, I can't reproduce the issues I found before any longer 🙏
   
   - Can't add more than one item: Fixed ✅
   - Existing array item values are reset to default props when adding an item: Fixed ✅
   - Field values are emptied when they lose focus in an array item: Fixed ✅
   - Duplicating items sets them to undefined: Fixed ✅
   - Moving items removes them from the page: Fixed ✅
   - Doing any changes in the top-level array sets nested array fields to undefined: Fixed ✅
   - Adding nested items sets the parent fields to undefined: Fixed ✅
   - Using the dispatcher to update the field (#1429): Fixed ✅
   - Using object fields at the root (#1450): Fixed ✅

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>
